File: 5-Problem-A-Day/165. Compare Version Numbers.java
class Solution {
    public int compareVersion(String version1, String version2) {
        int i = 0, j = 0, v1 = 0, v2 = 0;
        while (i < version1.length() && j < version2.length()) {
            char c1 = version1.charAt(i), c2 = version2.charAt(j);
            if (c1 == '.' && c2 == '.') {
                if (v1 > v2)
                    return 1;
                else if (v1 < v2)
                    return -1;
                v1 = v2 = 0;
                i++;
                j++;
            }
            if (c1 != '.') {
                v1 = v1 * 10 + Character.getNumericValue(c1);
                i++;
            }
            if (c2 != '.') {
                v2 = v2 * 10 + Character.getNumericValue(c2);
                j++;
            }
        }
        while (i < version1.length()) {
            char c1 = version1.charAt(i);
            if (c1 == '.') {
                if (v1 > v2)
                    return 1;
                if (v1 < v2)
                    return -1;
                v1 = v2 = 0;
            } else
                v1 = v1 * 10 + Character.getNumericValue(c1);
            i++;
        }
        while (j < version2.length()) {
            char c2 = version2.charAt(j);
            if (c2 == '.') {
                if (v1 > v2)
                    return 1;
                if (v1 < v2)
                    return -1;
                v1 = v2 = 0;
            } else
                v2 = v2 * 10 + Character.getNumericValue(c2);
            j++;
        }
        if (v1 > v2)
            return 1;
        if (v1 < v2)
            return -1;
        return 0;
    }
}

File: 2021/165. Compare Version Numbers.cpp
class Solution {
public:
    int compareVersion(string version1, string version2) {
        int i = 0, j = 0;
        while (i < version1.length() || j < version2.length()) {
            int v1 = 0, v2 = 0;
            if (i < version1.length()) {
                int dot = version1.find(".", i);
                if (dot == string::npos) {
                    v1 = stoi(version1.substr(i));
                    i = version1.length();
                } else {
                    v1 = stoi(version1.substr(i, dot - i));
                    i = dot + 1;
                }                
            }
            if (j < version2.length()) {
                int dot = version2.find(".", j);
                if (dot == string::npos) {
                    v2 = stoi(version2.substr(j));
                    j = version2.length();
                } else {                    
                    v2 = stoi(version2.substr(j, dot - j));
                    j = dot + 1;
                }                
                
            }
            if (v1 != v2) return v1 > v2 ? 1 : -1;
        }
        return 0;
    }
};